### PR TITLE
US08 Search Box

### PR DESCRIPTION
**Pull request by khanh000huynh**

### CODE REVIEW

File: pubspec.lock [hunk 1]
```diff
@@ -282,6 +282,13 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "2.0.1"
+  http:
```
Comment (TDKBao) [inline pubspec.lock]: Do not add pubspec.lock to repo

File: lib/src/widgets/search/search_bar.dart [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        padding: dropdownPadding,
+        child: Text(
+          searchState.failureMessage,
+          style: TextStyle(color: AppColor.secondaryGrey),
+        ),
+      );
+    }
+  }
+
+  List<Map<String, String>> getFormattedResult(
```
Comment (TDKBao) [inline lib/src/widgets/search/search_bar.dart]: This only hightlight 1 word matching right ?  Mr.Thanh require like page . It's mean every single word matching need hightlight not 1 . https://www.edamam.com/results/recipes/?search=chicken 

image

File: test/unit_tests/bloc_tests/search_bloc_test.dart [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  tearDown(() {
+    searchBloc.close();
+  });
+
+  blocTest('emits [] when no event is called',
+      build: () => searchBloc, expect: () => []);
+
+  blocTest(
+      'emits [SearchTextFieldChangeSuccess] when [SearchTextFieldChanged] is called and text field is empty',
```
Comment (TDKBao) [inline test/unit_tests/bloc_tests/search_bloc_test.dart]: Why logic emits success but text field empty ? When success it's mean already data right ?

Comment (khanh000huynh) [reply test/unit_tests/bloc_tests/search_bloc_test.dart]: This state is **SearchTextFieldChangeSuccess** which means whenever user stop typing, it receive a new value of the text filed. Just imagine like you type in 'chicken', it update **recipeTextFieldValue** to be 'chicken' and if you delete them all later, the new value of course will be empty. Maybe you misunderstood this state with **SearchRecipeSuccess(recipes: [])**. The test case you see here makes sure when the text field is empty then just update the **recipeTextFieldValue** and must not do anything else (search request). It is the opposite test case of the underneath one.

Comment (TDKBao) [reply test/unit_tests/bloc_tests/search_bloc_test.dart]: okay, I got that, you can skip it.

**Review by thanhphamtrung — COMMENTED** [review]
assets/images/UxRhrU8fPHQ.jpg => should rename this image to easily know what it is.

File: lib/src/blocs/search_bloc/search_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import 'dart:async';
```
Comment (thanhphamtrung) [inline lib/src/blocs/search_bloc/search_bloc.dart]: search_bloc.dart =>keyword_search_bloc.dart

File: lib/src/services/search_services.dart [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import 'package:http/http.dart' as http;
+
+import '../constants/constant_text.dart';
+import '../models/recipe_model.dart';
+
+class SearchServices {
+  Future<List<RecipeModel>> searcRecipesByName(String searchQuery) async {
+    final _recipesUrl =
+        'https://api.edamam.com/api/recipes/v2?type=public&q=$searchQuery&app_id=7024d4bb&app_key=cc06283adc212290729acaf0fe2520d9';
```
Comment (thanhphamtrung) [inline lib/src/services/search_services.dart:11]: Should you split this query to ez to maintain

File: lib/src/widgets/search/search_bar.dart [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+  BorderRadius getSearchTextFieldBorderRadius() {
+    return dropdownOverlayEntry != null
+        ? BorderRadius.only(
+            topLeft: searchTextFieldAndDropdownRadius,
+            topRight: Radius.circular(8),
+          )
+        : BorderRadius.circular(8);
+  }
+
+  void openDropdown(BuildContext context) async {
```
Comment (thanhphamtrung) [inline lib/src/widgets/search/search_bar.dart:259]: openDropdown => _openDropdown

File: lib/src/widgets/search/search_bar.dart [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+                      bottomRight: searchTextFieldAndDropdownRadius,
+                    )),
+                child: Material(
+                    color: Colors.transparent, child: getDropdownWidget())),
+          ));
+    });
+    Overlay.of(context)?.insert(dropdownOverlayEntry!);
+  }
+
+  void closeDropdown() {
```
Comment (thanhphamtrung) [inline lib/src/widgets/search/search_bar.dart:288]: closeDropdown => _closeDropdown

File: lib/src/widgets/search/search_bar.dart [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+    if (dropdownOverlayEntry != null) {
+      dropdownOverlayEntry!.remove();
+      dropdownOverlayEntry = null;
+      if (searchTextEditingController.text.isEmpty) {
+        recipesByName = [];
+      }
+    }
+  }
+
+  void textFieldOnChangedCallback(String text) {
```
Comment (thanhphamtrung) [inline lib/src/widgets/search/search_bar.dart]: textFieldOnChangedCallback => _textFieldOnChangedCallback

File: lib/src/widgets/search/search_bar.dart [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    const _duration = Duration(milliseconds: 500);
+    final _searchBloc = context.read<SearchBloc>();
+    final VoidCallback handleChange = () {
+      _searchBloc.add(SearchTextFieldChanged(recipeTextFieldValue: text));
+    };
+    _searchBloc.add(SearchInitialReturned());
```
Comment (thanhphamtrung) [inline lib/src/widgets/search/search_bar.dart]: SearchInitialReturned why we need initial here?

File: test/unit_tests/bloc_tests/search_bloc_test.dart [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  Future<List<RecipeModel>> searcRecipesByName(String searchQuery) async {
+    if (searchQuery == fakeValidSearchValue) {
+      return fakeRecipes;
+    }
+    if (searchQuery == fakeNoResultValue) {
+      return [];
+    }
+    throw Exception();
+  }
+}
```
Comment (thanhphamtrung) [inline test/unit_tests/bloc_tests/search_bloc_test.dart:32]: can I use when(searchServices.searcRecipesByName).thenAnswer()??

File: test/unit_tests/bloc_tests/search_bloc_test.dart (hunk 3 above)
Comment (thanhphamtrung) [reply test/unit_tests/bloc_tests/search_bloc_test.dart]: so do we need to call api with empty query?

File: lib/src/services/search_services.dart (hunk 5 above)
Comment (khanh000huynh) [reply lib/src/services/search_services.dart:11]: Could you be more specific? What should I do with this url?

File: test/unit_tests/bloc_tests/search_bloc_test.dart (hunk 3 above)
Comment (khanh000huynh) [reply test/unit_tests/bloc_tests/search_bloc_test.dart]: No, I never did that. If the text field is empty, then **no** search request will be sent at all.

File: pubspec.lock (hunk 1 above)
Comment (khanh000huynh) [reply pubspec.lock]: Done

**Pull request merged**